### PR TITLE
New gt sort

### PR DESCRIPTION
**Pull request by permcody**
A little cleanup with more C++11. However this also does a better sort on the initial id assignment for the grains. It shouldn't depend on the number of processors, so we are sorting on the minimum mesh entity in the grain.

refs #6711

### CODE REVIEW

**Comment by permcody** [comment]
Contains gratuitous and possibly unnecessary use of standard algorithms!

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h [hunk 1]
```diff
@@ -91,18 +91,13 @@ class FeatureFloodCount :
...
 
-    FeatureData(const FeatureData & f) :
-        _ghosted_ids(f._ghosted_ids),
-        _local_ids(f._local_ids),
-        _halo_ids(f._halo_ids),
-        _periodic_nodes(f._periodic_nodes),
-        _var_idx(f._var_idx),
-        _bboxes(f._bboxes),
-        _min_entity_id(f._min_entity_id),
-        _status(NOT_MARKED),
```
Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: This was different from the default copy ctor. Just an observation.

Comment (permcody) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Wow - good catch! I'd better make sure that this isn't a problem before this is merged.

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 2]
```diff
@@ -308,15 +308,30 @@ GrainTracker::trackGrains()
...
       for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-        for (unsigned int feature_num = 0; feature_num < _feature_sets[map_num].size(); ++feature_num)
-        {
-          _feature_sets[map_num][feature_num]->_status = MARKED;
-          _unique_grains[counter++] = std::move(_feature_sets[map_num][feature_num]);
-        }
+      {
+        // Sort the grains represented by this variable by _min_entity_id
+        std::sort(_feature_sets[map_num].begin(), _feature_sets[map_num].end(),
+                  [](const std::unique_ptr<FeatureData> & lhs, const std::unique_ptr<FeatureData> & rhs)
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I like it!

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 3]
```diff
@@ -308,15 +308,30 @@ GrainTracker::trackGrains()
...
+                  {
+                    return lhs->_min_entity_id < rhs->_min_entity_id;
+                  });
+
+        // Transfer the grains to the _unique_grains structure and assign ids
+        std::transform(_feature_sets[map_num].begin(), _feature_sets[map_num].end(), std::inserter(_unique_grains, _unique_grains.end()),
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: This one I can't parse!  Doesn't mean it's bad though!

Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I'm turning an array that looks like this { A, D, C, B, F, E}
into a map like this { (1,A), (2,D), (3,D), (4, B), (5, F), (6, E) }

Yeah this could be trivially turned into a `for` loop. Perhaps that would be better...

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h (hunk 1 above)
Comment (permcody) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: OK, this is fine. I don't do any copying at all now (to my knowledge), and even if I do the `_status` variable wouldn't be expected to be preserved or used at all. In fact, I'm somewhat tempted to `delete` both the copy constructor and copy assignment operator.... I wonder if this is good practice?

Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: I would give that a shot. If you are not 100% sure the copy ctor does the right thing you might get bitten by accidental copying. Deleting the ctors will avoid that by throwing compile time errors.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 3 above)
Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: No, leave it, it is very elegant and a good example. I would add `{A,B,C}->{(0,A),(1,B),(2,C)}` to the comment above it though.

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 4]
```diff
@@ -308,15 +308,30 @@ GrainTracker::trackGrains()
...
-    else // Just assign IDs in order
+    else
     {
+      /**
+       * Here we want to assign the grains in some partitioning invariant way. We'll sort first by
+       * _var_idx (already partitioned in that manner) then sort on the _min_entity_id
+       * to order the grains.
+       */
       unsigned int counter = 0;
       for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I strongly suggest changing this to

```
for (decltype(_maps_size) map_num = 0; map_num < _maps_size; ++map_num)
```

The detour using auto and the somewhat cryptic `decltype(_maps_size)(0)` looks hard to parse for newbies (i.e. regular humans).

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 3 above)
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: >I'm turning an array that looks like this { A, D, C, B, F, E}
into a map like this { (1,A), (2,D), (3,D), (4, B), (5, F), (6, E) }

Makes sense - maybe just add that as a comment?

This is a cool thing to do BTW... Python even has this built in (`enumerate()`: https://docs.python.org/2/library/functions.html#enumerate )

It might be worth actually creating a little `MooseUtils` function / macro for this.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 4 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Yeah I had it like that, before. I think the consensus is that this syntax is "gross".

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h (hunk 1 above)
Comment (friedmud) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: I like the idea of explicitly deleting the copy constructor if you don't ever want the object to be copied (or it's unsafe to copy it)

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 4 above)
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I like `dschwen`'s syntax better.

I'm unclear on which one @permcody is referring when he says:

>this syntax is "gross"

Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I looked around online and think people have a hard time seeing what looks like a function call as the type. This is in contract to having auto as the type and just looking at the expression where it comes from. I could change it, but I used the latter syntax everywhere else in FeatureFloodCount and GrainTracker so I've committed to the latter.

Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Interesting.  I really don't like `auto map_num = decltype(_maps_size)(0)`.  It goes against our guidelines that say not to use constructor `()` syntax for simple objects!  And if this wasn't a simple object then it would go against our guidelines to not use assignment construction syntax for complicated objects.

So, it's a double whammy on our style guidelines.

`decltype(_maps_size) map_num = 0` is very easy to parse.  In my mind `decltype()` is just like a preprocessor directive / macro... and my mind very happily just swaps in the type of `_maps_type` right into the code there so it looks like a normal `for` loop.

Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Fine with me.

Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Oh - well now I remember part of the reason I didn't go with that syntax. It's because the actual type for many of these counters is const! `_maps_size` is `const unsigned int`. Therefore the code won't compile without a really ugly const cast! I searched for awhile and the syntax I have now was one of the better ones I encountered.

Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Oh wow - that really sucks!

It's kind of weird that it's `const` on the LHS... but not on the RHS... what does that RHS expand to?

Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]:
```
for (std::remove_const<decltype(_maps_size)>::type map_num = 0; map_num < _maps_size; ++map_num)
```

;-) ok, maybe not...

Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: lol - yeah... I think @permcody must have the right answer here even though I'm not happy about it. 😄

**Comment by permcody** [comment]
OK @friedmud and @dschwen, this PR just got a whole lot cooler! I've managed to strip out all of the indirect container referencing. I'm now storing the "FeatureData" objects directly in standard containers instead of through pointers. To ensure that I don't have any copying going on (which is how I found the extra copy happing in the Console), I've explicitly deleted the copy ctor and copy assignment operator for that class.

It really cleans things up. No more arrow operators, and I managed to drop two of my recently added sort lambdas since I have a less-than operator defined for this struct. This code should be more efficient than ever since nothing is ever copied!

**Comment by dschwen** [comment]
Yeah, this looks elegant. Not so sure about the _faster than ever_ though, since moving is not for free. I expect a lot more data to get shuffled around when sorting a container of objects rather than pointers. But this is probably not a significant cost anyways.

**Comment by permcody** [comment]
Hmm. I guess that's true, not data but essentially several allocation
pointers.

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 5]
```diff
@@ -118,12 +118,12 @@ GrainTracker::finalize()
...
     {
-      if (grain_pair.second->_status != INACTIVE)
+      if (grain_pair.second._status != INACTIVE)
       {
-        for (auto elem_id : grain_pair.second->_local_ids)
+        for (auto elem_id : grain_pair.second._local_ids)
         {
           mooseAssert(!_ebsd_reader || _unique_grain_to_ebsd_num.find(grain_pair.first) != _unique_grain_to_ebsd_num.end(), "Bad mapping in unique_grain_to_ebsd_num");
-          _elemental_data[elem_id].push_back(std::make_pair(_ebsd_reader ? _unique_grain_to_ebsd_num[grain_pair.first] : grain_pair.first, grain_pair.second->_var_idx));
+          _elemental_data[elem_id].push_back(std::make_pair(_ebsd_reader ? _unique_grain_to_ebsd_num[grain_pair.first] : grain_pair.first, grain_pair.second._var_idx));
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I think it would be more efficient to use `emplace_back()` here amd pass in the two arguments for the pair.

**Comment by friedmud** [comment]
This looks great!

As for speed... it doesn't matter.  This is what modern C++ is supposed to look like.  We're supposed to be going away from using pointers to store objects in containers and towards using move semantics to store objects directly.

I say: "well done!"

:clap:

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 5 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I don't think it's more efficient but it might look better. Note that I
deleted the copy ctors so this code is doing the right thing.

Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: `emplace_back()` with passing in the two arguments will be more efficient.

What you have now creates two pairs, then calls the move constructor then destroys one of the pairs.

Using `emplace_back()` it will simply create one pair in the vector, at the right spot in memory.  Done.

Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Ok, I see

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 6]
```diff
@@ -417,13 +432,17 @@ GrainTracker::trackGrains()
...
    */
   for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-    for (unsigned int i = 0; i < _feature_sets[map_num].size(); ++i)
-      if (_feature_sets[map_num][i] && _feature_sets[map_num][i]->_status == NOT_MARKED)
+    for (auto feature_num = decltype(_feature_sets[map_num].size())(0);
+         feature_num < _feature_sets[map_num].size(); ++feature_num)
+      // If it's not in the index list, it hasn't been transferred
+      if (new_grain_idx_to_existing_grain_idx.find(std::make_pair(map_num, feature_num)) == new_grain_idx_to_existing_grain_idx.end())
       {
+        mooseAssert(_feature_sets[map_num][i]._status == NOT_MARKED, "Feature in wrong state, logic error");
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: This line causes the compile error in debug. `i` is the index of the old for loop you removed. Needs to be `feature_num`.

Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Damn. I really need to switch to "devel" mode

**Comment by permcody** [comment]
This should be ready to merge
